File: src/evai/command/ScanCommands.java
package evai.command;

import mindustry.*;
import mindustry.entities.*;
import mindustry.gen.*;
import mindustry.game.*;
import mindustry.world.*;
import mindustry.world.blocks.environment.*;
import java.util.*;

public class ScanCommands implements CommandRegistry.CommandHandler {
    @Override
    public CommandResult handle(String action, String[] args) {
        switch (action) {
            case "scan": return scan(args);
            case "check": return check(args);
            case "map": {
                Map<String, Object> d = new HashMap<>();
                d.put("width", Vars.world.width());
                d.put("height", Vars.world.height());
                d.put("map", Vars.state.map != null ? Vars.state.map.name() : "none");
                d.put("wave", (int) Vars.state.wave);
                return new CommandResult(true, "", d);
            }
            default: return new CommandResult(false, "unknown scan command: " + action);
        }
    }

    private CommandResult scan(String[] args) {
        Player player = Vars.player;
        if (player == null) return new CommandResult(false, "no player");
        if (args.length > 0 && (args[0].equals("resources") || args[0].equals("liquids"))) {
            return scanResources(player, args);
        }
        int radius = args.length > 0 ? tryParseInt(args[0], 20) : 20;
        int tx = player.tileX(), ty = player.tileY();
        List<Map<String, Object>> blocks = new ArrayList<>();
        int minX = Math.max(tx - radius, 0);
        int maxX = Math.min(tx + radius, Vars.world.width() - 1);
        int minY = Math.max(ty - radius, 0);
        int maxY = Math.min(ty + radius, Vars.world.height() - 1);
        for (int x = minX; x <= maxX; x++) {
            for (int y = minY; y <= maxY; y++) {
                Tile tile = Vars.world.tile(x, y);
                if (tile == null) continue;
                Block blk = tile.block();
                if (blk == null || blk instanceof AirBlock || blk instanceof Floor || blk instanceof OverlayFloor) continue;
                Building bld = tile.build;
                Map<String, Object> m = new HashMap<>();
                m.put("x", x); m.put("y", y);
                m.put("block", blk.name);
                m.put("team", tile.team() != null ? tile.team().name : "none");
                m.put("health", bld != null ? (int) bld.health() : 0);
                blocks.add(m);
            }
        }
        Map<String, Object> data = new HashMap<>();
        Map<String, Object> center = new HashMap<>();
        center.put("x", tx); center.put("y", ty);
        data.put("center", center);
        data.put("radius", radius);
        data.put("count", blocks.size());
        data.put("blocks", blocks);
        return new CommandResult(true, "", data);
    }

    private CommandResult scanResources(Player player, String[] args) {
        int radius = args.length > 1 ? tryParseInt(args[1], 20) : 20;
        int tx = player.tileX(), ty = player.tileY();
        List<Map<String, Object>> resources = new ArrayList<>();
        int minX = Math.max(tx - radius, 0);
        int maxX = Math.min(tx + radius, Vars.world.width() - 1);
        int minY = Math.max(ty - radius, 0);
        int maxY = Math.min(ty + radius, Vars.world.height() - 1);
        for (int x = minX; x <= maxX; x++) {
            for (int y = minY; y <= maxY; y++) {
                Tile tile = Vars.world.tile(x, y);
                if (tile == null) continue;
                if (args[0].equals("resources")) {
                    Block ov = tile.overlay();
                    if (ov != null && !(ov instanceof AirBlock) && ov.itemDrop != null) {
                        Map<String, Object> m = new HashMap<>();
                        m.put("x", x); m.put("y", y);
                        m.put("ore", ov.name);
                        m.put("item", ov.itemDrop.name);
                        resources.add(m);
                    }
                } else {
                    Floor fl = tile.floor();
                    if (fl != null && fl.liquidDrop != null) {
                        Map<String, Object> m = new HashMap<>();
                        m.put("x", x); m.put("y", y);
                        m.put("liquid", fl.liquidDrop.name);
                        resources.add(m);
                    }
                }
            }
        }
        Map<String, Object> data = new HashMap<>();
        data.put("type", args[0]);
        data.put("count", resources.size());
        data.put("resources", resources);
        return new CommandResult(true, "", data);
    }

    private CommandResult check(String[] args) {
        Player player = Vars.player;
        if (player == null) return new CommandResult(false, "no player");
        int idx = 0;
        String field = "";
        if (args.length >= 3 && (args[0].equals("items") || args[0].equals("power") || args[0].equals("liquid")
            || args[0].equals("health") || args[0].equals("tile") || args[0].equals("enemies") || args[0].equals("allies"))) {
            field = args[0]; idx = 1;
        }
        if (field.equals("enemies") || field.equals("allies")) {
            int radius = args.length > idx ? tryParseInt(args[idx], 50) : 50;
            Team playerTeam = player.team();
            final String targetField = field;
            List<Map<String, Object>> nearby = new ArrayList<>();
            Units.nearby(player.x, player.y, radius * 2f, radius * 2f, u -> {
                boolean isAlly = u.team() == playerTeam;
                if ((targetField.equals("allies") && isAlly) || (targetField.equals("enemies") && !isAlly)) {
                    Map<String, Object> m = new HashMap<>();
                    m.put("name", u.type().name);
                    m.put("health", (int) u.health() + "/" + (int) u.maxHealth());
                    m.put("team", u.team().name);
                    m.put("x", (int) u.x());
                    m.put("y", (int) u.y());
                    nearby.add(m);
                }
            });
            Map<String, Object> data = new HashMap<>();
            data.put("type", field);
            data.put("count", nearby.size());
            data.put("units", nearby);
            return new CommandResult(true, "", data);
        }
        Integer x = args.length > idx ? tryParseInt(args[idx], null) : null;
        Integer y = args.length > idx + 1 ? tryParseInt(args[idx + 1], null) : null;
        if (x == null || y == null) return new CommandResult(false, "usage: check [field] <x> <y>");
        Tile tile = Vars.world.tile(x, y);
        if (tile == null) return new CommandResult(false, "tile not found");
        Building bld = tile.build;

        if (field.equals("tile")) {
            Map<String, Object> d = new HashMap<>();
            d.put("x", tile.x); d.put("y", tile.y);
            d.put("block", tile.block() != null ? tile.block().name : "air");
            d.put("floor", tile.floor() != null ? tile.floor().name : "?");
            d.put("overlay", tile.overlay() != null ? tile.overlay().name : "none");
            d.put("team", tile.team() != null ? tile.team().name : "none");
            d.put("solid", tile.solid());
            return new CommandResult(true, "", d);
        } else if (field.equals("items")) {
            if (bld == null) return new CommandResult(false, "no building");
            Map<String, Object> d = new HashMap<>();
            d.put("items", bld.items != null ? bld.items.total() : 0);
            d.put("capacity", bld.block.itemCapacity);
            return new CommandResult(true, "", d);
        } else if (field.equals("power")) {
            if (bld == null) return new CommandResult(false, "no building");
            Map<String, Object> d = new HashMap<>();
            d.put("status", bld.power != null ? bld.power.status : 0.0);
            d.put("produced", bld.power != null && bld.power.graph != null ? bld.power.graph.getPowerProduced() : 0.0);
            d.put("needed", bld.power != null && bld.power.graph != null ? bld.power.graph.getPowerNeeded() : 0.0);
            return new CommandResult(true, "", d);
        } else if (field.equals("liquid")) {
            if (bld == null) return new CommandResult(false, "no building");
            float liquidTotal = 0f;
            if (bld.liquids != null) {
                for (int li = 0; li < Vars.content.liquids().size; li++) {
                    liquidTotal += bld.liquids.get(Vars.content.liquid(li));
                }
            }
            Map<String, Object> d = new HashMap<>();
            d.put("liquids", liquidTotal);
            d.put("capacity", bld.block.liquidCapacity);
            return new CommandResult(true, "", d);
        } else if (field.equals("health")) {
            if (bld == null) return new CommandResult(false, "no building");
            Map<String, Object> d = new HashMap<>();
            d.put("health", (int) bld.health() + "/" + (int) bld.maxHealth());
            return new CommandResult(true, "", d);
        } else {
            Map<String, Object> d = new HashMap<>();
            d.put("x", tile.x); d.put("y", tile.y);
            d.put("block", tile.block() != null ? tile.block().name : "air");
            d.put("floor", tile.floor() != null ? tile.floor().name : "?");
            d.put("team", tile.team() != null ? tile.team().name : "none");
            if (bld != null) {
                Map<String, Object> bd = new HashMap<>();
                bd.put("health", (int) bld.health() + "/" + (int) bld.maxHealth());
                bd.put("enabled", bld.enabled);
                bd.put("warmup", bld.warmup());
                bd.put("efficiency", bld.efficiency);
                bd.put("rotation", bld.rotation);
                bd.put("items", bld.items != null ? bld.items.total() : 0);
                bd.put("itemCapacity", bld.block.itemCapacity);
                d.put("building", bd);
            }
            return new CommandResult(true, "", d);
        }
    }

    private static int tryParseInt(String s, int fallback) {
        try { return Integer.parseInt(s); } catch (NumberFormatException e) { return fallback; }
    }

    private static Integer tryParseInt(String s, Integer fallback) {
        try { return Integer.parseInt(s); } catch (NumberFormatException e) { return fallback; }
    }
}

File: src/evai/command/ScanCommands.kt
package evai.command

import mindustry.*
import mindustry.entities.*
import mindustry.game.*
import evai.command.CommandRegistry.CommandHandler

class ScanCommands : CommandHandler {
    override fun handle(action: String, args: Array<String>): CommandResult = when (action) {
        "scan" -> {
            if (args.isEmpty() || args[0] == "resources" || args[0] == "liquids") {
                scanResources(args)
            } else {
                val radius = args[0].toIntOrNull() ?: 20
                val player = Vars.player ?: return CommandResult(false, "no player")
                val tx = player.tileX()
                val ty = player.tileY()
                val blocks = mutableListOf<Map<String, Any?>>()
                for (x in (tx - radius).coerceAtLeast(0)..(tx + radius).coerceAtMost(Vars.world.width() - 1)) {
                    for (y in (ty - radius).coerceAtLeast(0)..(ty + radius).coerceAtMost(Vars.world.height() - 1)) {
                        val tile = Vars.world.tile(x, y) ?: continue
                        if (tile.block() !== mindustry.world.blocks.environment.Air) {
                            val build = tile.build()
                            blocks.add(mapOf(
                                "x" to x, "y" to y,
                                "block" to (tile.block()?.name ?: "?"),
                                "team" to (tile.team()?.name ?: "none"),
                                "health" to (build?.health?.toInt() ?: 0)
                            ))
                        }
                    }
                }
                CommandResult(true, data = mapOf("center" to mapOf("x" to tx, "y" to ty), "radius" to radius, "count" to blocks.size, "blocks" to blocks))
            }
        }
        "check" -> {
            var idx = 0
            var field = ""
            if (args.size >= 3 && args[0] in listOf("items", "power", "liquid", "health", "tile", "enemies", "allies")) {
                field = args[0]
                idx = 1
            }
            if (field == "enemies" || field == "allies") {
                val radius = if (args.size > idx) args[idx].toIntOrNull() ?: 50 else 50
                val player = Vars.player ?: return CommandResult(false, "no player")
                val team = if (field == "enemies") player.team().enemies else player.team()
                @Suppress("UNCHECKED_CAST")
                val units = (arc.struct.Seq::class.java.getMethod("copy", Unit::class.java) as? java.lang.reflect.Method)?.let {
                    // Try to get nearby units
                    val nearby = mutableListOf<Map<String, Any?>>()
                    Units.nearby(player.x, player.y, radius.toFloat()) { u ->
                        if (u.team == team) {
                            nearby.add(mapOf("name" to (u.type?.name ?: "?"), "health" to "${u.health.toInt()}/${u.maxHealth.toInt()}", "x" to u.x.toInt(), "y" to u.y.toInt()))
                        }
                    }
                    nearby
                } ?: run {
                    val nearby = mutableListOf<Map<String, Any?>>()
                    Units.nearby(player.x, player.y, radius.toFloat()) { u ->
                        if (u.team == team) {
                            nearby.add(mapOf("name" to (u.type?.name ?: "?"), "health" to "${u.health.toInt()}/${u.maxHealth.toInt()}", "x" to u.x.toInt(), "y" to u.y.toInt()))
                        }
                    }
                    nearby
                }
                CommandResult(true, data = mapOf("type" to field, "count" to units.size, "units" to units))
            } else {
                val x = if (args.size > idx) args[idx].toIntOrNull() else null
                val y = if (args.size > idx + 1) args[idx + 1].toIntOrNull() else null
                if (x == null || y == null) return CommandResult(false, "usage: check [field] <x> <y>")
                val tile = Vars.world.tile(x, y) ?: return CommandResult(false, "tile not found")
                when (field) {
                    "tile" -> CommandResult(true, data = mapOf("x" to tile.x, "y" to tile.y, "block" to (tile.block()?.name ?: "air"), "floor" to (tile.floor()?.name ?: "?"), "overlay" to (tile.overlay()?.name ?: "none"), "team" to (tile.team()?.name ?: "none"), "solid" to tile.solid()))
                    "items" -> {
                        val build = tile.build() ?: return CommandResult(false, "no building")
                        CommandResult(true, data = mapOf("items" to (if (build.items != null) build.items.total() else 0), "capacity" to build.block.itemCapacity))
                    }
                    "power" -> {
                        val build = tile.build() ?: return CommandResult(false, "no building")
                        val power = build.power
                        CommandResult(true, data = mapOf("status" to (power?.status ?: "none"), "produced" to (power?.graph?.getPowerProduced() ?: 0.0), "needed" to (power?.graph?.getPowerNeeded() ?: 0.0)))
                    }
                    "liquid" -> {
                        val build = tile.build() ?: return CommandResult(false, "no building")
                        CommandResult(true, data = mapOf("liquids" to (build.liquids?.total() ?: 0.0), "capacity" to (build.block.liquidCapacity)))
                    }
                    "health" -> {
                        val build = tile.build() ?: return CommandResult(false, "no building")
                        CommandResult(true, data = mapOf("health" to "${build.health}/${build.maxHealth}"))
                    }
                    "" -> {
                        val build = tile.build()
                        val data = mutableMapOf<String, Any?>("x" to tile.x, "y" to tile.y, "block" to (tile.block()?.name ?: "air"), "floor" to (tile.floor()?.name ?: "?"), "team" to (tile.team()?.name ?: "none"))
                        if (build != null) {
                            data["building"] = mapOf("health" to "${build.health}/${build.maxHealth}", "enabled" to build.enabled, "warmup" to build.warmup, "efficiency" to build.efficiency, "rotation" to build.rotation, "items" to (build.items?.total() ?: 0), "itemCapacity" to build.block.itemCapacity)
                        }
                        CommandResult(true, data = data)
                    }
                    else -> CommandResult(false, "unknown field: $field")
                }
            }
        }
        "map" -> {
            val world = Vars.world
            val state = Vars.state
            CommandResult(true, data = mapOf("width" to world.width(), "height" to world.height(), "map" to (state.map?.name() ?: "none"), "wave" to state.wave))
        }
        else -> CommandResult(false, "unknown scan command: $action")
    }

    private fun scanResources(args: Array<String>): CommandResult {
        val player = Vars.player ?: return CommandResult(false, "no player")
        val radius = if (args.size > 1) args[1].toIntOrNull() ?: 20 else 20
        val tx = player.tileX()
        val ty = player.tileY()
        val resources = mutableListOf<Map<String, Any?>>()
        for (x in (tx - radius).coerceAtLeast(0)..(tx + radius).coerceAtMost(Vars.world.width() - 1)) {
            for (y in (ty - radius).coerceAtLeast(0)..(ty + radius).coerceAtMost(Vars.world.height() - 1)) {
                val tile = Vars.world.tile(x, y) ?: continue
                if (args[0] == "resources" && tile.overlay()?.itemDrop != null) {
                    resources.add(mapOf("x" to x, "y" to y, "ore" to tile.overlay()?.name, "item" to tile.overlay()?.itemDrop?.name))
                } else if (args[0] == "liquids" && tile.floor()?.liquidDrop != null) {
                    resources.add(mapOf("x" to x, "y" to y, "liquid" to tile.floor()?.liquidDrop?.name))
                }
            }
        }
        return CommandResult(true, data = mapOf("type" to args[0], "count" to resources.size, "resources" to resources))
    }
}